### ISSUE TITLE
[BUG] No error detection in corrupted ORC files

### ISSUE THREAD

**Issue by revans2** [rapidsai/cudf#13461]
**Describe the bug**
This is related to https://github.com/rapidsai/cudf/issues/13460

After we write bad data, when we read it back in on the GPU we don't produce any errors.  We don't crash. We just happily return corrupt data.

```
scala> spark.time(spark.read.parquet("./target/TMP_PAR").selectExpr("MIN(ts)", "MAX(ts)").show(false))
+--------------------------+--------------------------+
|min(ts)                   |max(ts)                   |
+--------------------------+--------------------------+
|2023-05-23 08:34:20.007655|2023-05-23 08:34:20.993544|
+--------------------------+--------------------------+

Time taken: 189 ms

scala> spark.time(spark.read.orc("./target/TMP_ORC").selectExpr("MIN(ts)", "MAX(ts)").show(false))
+-------------------+-------------------+
|min(ts)            |max(ts)            |
+-------------------+-------------------+
|2015-01-01 00:00:00|2015-01-01 00:00:08|
+-------------------+-------------------+
```

I realize that CUDF for performance reasons does not do much in the way of checks on the input data, but for input files we really should be doing something.

**Comment by GregoryKimball** [comment]
We previously added support for "error code" tracking in the cuDF Parquet reader, starting with #14167 and follow-on work in #14237 and #14706. We should expand this pattern to manage decoding errors in the cuDF ORC reader as well.

**Comment by vuule** [comment]
ORC reader does not have an error code, but it does have an error counter and throws if it's non-zero.

However, this type of corruption is not caught by the kernel. I looked into this and it turns out it's pretty hard to detect.

What happens in the kernel is that the corrupted data **fails to decode** (i.e. there are zero decoded row in the row group). This is not taken into account when copying the elements to the output column, so garbage is written. I have [a small change](https://github.com/vuule/cudf/tree/bug-orc-detect-failed-decode) that prevent writing garbage to the output, but we still end up with zeroes in the column.

One way to detect this corruption would be to compare the number of decoded elements in the row group to the valid row count at the end of the kernel. But, AFAICT, we don't have the number of valid rows in each row group available and calculating this would take extra time.

We do have a valid row count for each stripe, and I implemented a counter for the number of decoded rows for the whole stripe. In theory, mismatch between the two value should indicate failed decode. But, these values don't match in a few cases, worst of which is when `num_rows`/`skip_rows` are used.

Asked @kingcrimsontianyu to check my conclusions here, but there's no ETA on we we'll be able to look into this further.

TL;DR We can add a decode check that's enabled in many (most?) cases, and adds very little overhead (single atomicAdd per row group). Not yet sure if we can do better.
